### ISSUE TITLE
[QST] How do I import the `cutlass::gemm::kernel::GemmUniversal`

### ISSUE THREAD

**Issue by mrakgr** [NVIDIA/cutlass#1814]
Following up on #1291, I am finally making the time to properly understand Cutlass. I can build and run the examples in the repo, but I am having a lot of difficulty understanding the type errors I am getting when I try to import the `GemmUniversal` classes. For example...

```cpp
#include <iostream>

#include "cutlass/cutlass.h"
#include "cute/tensor.hpp"
#include "cute/atom/mma_atom.hpp"

#include "cutlass/numeric_types.h"

#include "cutlass/gemm/device/gemm_universal_adapter.h"
#include "default_gemm_configuration.hpp"

using namespace cute;

/////////////////////////////////////////////////////////////////////////////////////////////////

using Config = cutlass::gemm::device::DefaultGemmConfigurationToCutlass3Types<
	cutlass::arch::OpClassSimt, cutlass::arch::Sm50,
	double, cutlass::layout::ColumnMajor,
	double, cutlass::layout::ColumnMajor,
	double, cutlass::layout::ColumnMajor,
	double>;

using GemmKernel = cutlass::gemm::kernel::GemmUniversal<
	Shape<int, int, int, int>,
	Config::CollectiveMainloop,
	Config::CollectiveEpilogue
>;

using Gemm = cutlass::gemm::device::GemmUniversalAdapter<GemmKernel>;
```

I've isolated this piece of code from one of the tests and am trying to bring it into another project without success. Even when I seemingly have all the necessary includes, I see something like this...

![Image](https://github.com/user-attachments/assets/507ae9d2-6ab8-4895-98cc-b8524d982462)

Probably I shouldn't be trying to import test harness code, but the trouble is that there are no clean examples how to use the universal kernel.

https://github.com/NVIDIA/cutlass/blob/main/media/docs/quickstart.md#launching-a-gemm-kernel-using-cutlass-30-or-newer

The quickstart one is clean but doesn't typecheck for me.

There are some examples like the `48_hopper_warp_specialized_gemm` which do have use examples of it, but when I get rid of the device arch checking macros, I just get a ton of type errors. Those kinds of examples aren't usable.

I am going to try recompiling the Cutlass library for 9.0 devices just to see if something changes.

But otherwise, could I get some help to start me off with Cutlass? Thanks.

**Comment by mrakgr** [comment]
Of course, just `#include "cutlass/gemm/kernel/gemm_universal.h"` doesn't work for me. The IDE doesn't see the class inside the file.

**Comment by mrakgr** [comment]
I just recompiled Cutlass with 90 support. Here is how it looks like in the IDE.

![Image](https://github.com/user-attachments/assets/4238fee3-4d97-4d5f-a7c0-39b9d2933ade)

`GroupProblemShape` has a type error, and these kinds of issues aren't present in the non-90 examples. More problematically, everything below `#if defined(CUTLASS_ARCH_MMA_MODIFIABLE_TMA_SM90_SUPPORTED)` is grayed out.

If I were to remove those macro checks...

![Image](https://github.com/user-attachments/assets/b40db96f-ab84-44aa-ba5d-dda9e598f5a0)
![Image](https://github.com/user-attachments/assets/4d90544f-2faf-4c4e-8013-a2aa686d82b7)
![Image](https://github.com/user-attachments/assets/4c8ec4d5-d169-4783-9c26-5c1f106b7618)

Please have mercy on me here.

**Comment by d-k-b** [comment]
@mrakgr, just to understand your setup, are you using the CMake configuration flow built into the IDE? Or are you just using the IDE to view/edit the files but building from elsewhere like powershell? I ask because the CUTLASS device-side code relies on many definitions that are computed and generated in CMake. To ensure things work properly, I would start at least by copying an existing example and get that working with a custom kernel you need, and that example to produce a library that contains the kernel for you to import in your other larger project (or something along those lines).

**Comment by mrakgr** [comment]
My actual setup is that I am using my own language [Spiral](https://github.com/mrakgr/Tutorials/blob/master/spiral_mini_tutorial_for_ml_library_authors.md) which compiles to Python + Cuda. You can see some examples in that tutorial to get a sense of how it looks. My stance towards complex C++ build setups is that I don't like them and don't want anything to do with them if possible. And it's not possible because Cutlass is hoisting CMake upon me, leaving me no choice but to interact with it. I've only been using VS because I am on Windows and am trying to handle on the library, which would be a lot harder from Spiral directly.

> I ask because the CUTLASS device-side code relies on many definitions that are computed and generated in CMake.

If that is true, that might make Cutlass unusable for me. I have a fully fused ML library and all the matmults need to be called on the device directly instead of being called from host. I've done [my own](https://github.com/mrakgr/Spiral-s-ML-Library/blob/d5152a52c150f4e2a8dd76599be6a7ffa58c6d16/ml/matmul/kernel.spi), but it's not as good as the Cutlass one, so my goal is to replace it with that. I want to find a way to use Cutlass as a header-only library, but if that is impossible, I'll give up and wait for the next gen consumer NVidia cards to come out. They have warp group matrix multiplication instructions operating on shared memory, which I expect should make it easier to make a fully performant matmult kernel.

Even if it's somehow possible to get it to compile by fitting all the right compiler options, I have a limited complexity budget. I am willing to do one or two extra compile time options to set up the Cutlass library when including it in Spiral code, but absolutely nothing like those enormous CMake files that I see in its repo.

**Comment by mrakgr** [comment]
Also, I should mention one aspect of my setup. All the tensor/matrix dimensions, layouts and their strides are known at compile time, so in theory, Cutlass should have everything it needs to select the optimal kernel at compile time without the need for build options.

**Comment by mrakgr** [comment]
In the ML library, I have poker games, which are running directly on the GPU, which are calling the ML library functions. As the games are register hungry, I've been running one block per SM, and the ML library and the matmult would need to be performant under that condition.
